stablehlo/dialect/Base.h

```diff
@@ -73,16 +73,27 @@ std::pair<int64_t, int64_t> inferConcatenatedDimAndBound(int64_t leftSize,
                                                          int64_t leftBound,
                                                          int64_t rightBound);
 
-FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
+FailureOr<std::pair<int64_t, int64_t>> inferMostSpecificDimAndBound(
     Optional<Location> location, int64_t dim, int64_t leftSize,
     int64_t rightSize, int64_t leftBound, int64_t rightBound);
 
+FailureOr<std::pair<int64_t, int64_t>> inferLeastSpecificDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound);
+
+// Infer single least specific return type from inputTypes with support for
+// bounds. (Size, bound) of each dimension of the return type will be merged
+// from corresponding dimensions of every inputType by extracting the least
+// specific one. Return unranked tensor if any input is unranked.
+FailureOr<Type> inferLeastSpecificType(Optional<Location> location,
+                                       TypeRange inputTypes);
+
 // Infer single most specific return type from inputTypes with support for
 // bounds. (Size, bound) of each dimension of the return type will be merged
-// from corresponding dimensions of every inputType by merging them.
-LogicalResult inferMostSpecificType(Optional<Location> location,
-                                    TypeRange inputTypes,
-                                    SmallVectorImpl<Type> &inferredReturnTypes);
+// from corresponding dimensions of every inputType by extracting the most
+// specific one. Return unranked tensor if all inputs are unranked.
+FailureOr<Type> inferMostSpecificType(Optional<Location> location,
```

Comment:
we could return ShapedType from here to avoid cast at the callsite?

Reply:
Done.